src/freenet/node/diagnostics/threads/DefaultThreadDiagnostics.java

```diff
@@ -0,0 +1,151 @@
+/* This code is part of Freenet. It is distributed under the GNU General
+ * Public License, version 2 (or at your option any later version). See
+ * http://www.gnu.org/ for further details of the GPL. */
+package freenet.node.diagnostics.threads;
+
+import freenet.node.*;
+import freenet.node.diagnostics.*;
+import freenet.support.*;
+
+import java.lang.management.*;
+import java.util.*;
+import java.util.concurrent.atomic.*;
+
+/**
+ * Runnable thread to retrieve node thread's information and compiling it into
+ * an array of NodeThreadInfo objects.
+ */
+public class DefaultThreadDiagnostics implements Runnable, ThreadDiagnostics {
+    private final String name;
+    private final int monitorInterval;
+
+    private final NodeStats nodeStats;
+    private final Ticker ticker;
+
+    /** Sleep interval to calculate % CPU used by each thread */
+    private static final int MONITOR_INTERVAL = 1000;
+    private static final String MONITOR_THREAD_NAME = "NodeDiagnostics: thread monitor";
+
+    private final AtomicReference<List<NodeThreadInfo>> nodeThreadInfo = new AtomicReference<>(new ArrayList<>());
+
+    private final OperatingSystemMXBean operatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean();
+    private final ThreadMXBean threadMxBean               = ManagementFactory.getThreadMXBean();
+    private final RuntimeMXBean runtimeMxBean             = ManagementFactory.getRuntimeMXBean();
+
+    /**
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     * @param name Thread name
+     * @param monitorInterval Sleep intervals to retrieve CPU usage
+     */
+    public DefaultThreadDiagnostics(NodeStats nodeStats, Ticker ticker, String name, int monitorInterval) {
+        this.nodeStats = nodeStats;
+        this.ticker = ticker;
+        this.name = name;
+        this.monitorInterval = monitorInterval;
+    }
+
+    /*
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     */
+    public DefaultThreadDiagnostics(NodeStats nodeStats, Ticker ticker) {
+        this(nodeStats, ticker, MONITOR_THREAD_NAME, MONITOR_INTERVAL);
+    }
+
+    private void scheduleNext(int interval) {
+        ticker.queueTimedJob(
+            this,
+            name,
+            interval,
+            false,
+            true
+        );
+    }
+
+    public void start() {
+        scheduleNext(0);
+    }
+
+    private void scheduleNext() {
+        scheduleNext(monitorInterval);
+    }
+
+    /** Map<ThreadId, ThreadCpuTimeNs> */
+    private final Map<Long, Long> threadCPU = new HashMap<>();
+    private long initialUptime = -1;
+
+    @Override
+    public void run() {
```

Comment:
Hmm, no, this isn’t what I meant. This needs two intervals to get the data; I’d prefer to have each run create new data by diffing it with the last run, except on the first run where no data is created.

Reply:
I don't follow.

Reply:
First run gathers initial data, then schedules the next run. Second run then creates the final data and resets `initialUptime` and `threadCPU`. This means that with an interval of 10 seconds the data only changes every 20 seconds.

All runs should do the same thing; get the current threads, get their previous CPU times, store the diff in the result. This way the data will never be older than the interval (ignoring processing time).

Reply:
Oh, I got you now. :+1: I'll adapt it to do that.

Reply:
Should be done now! Looks way better :)

Reply:
Okay, the basic loop is nailed down, this looks very good. 😄

I’m not too sure about the `totalCpuTime`, though… if enough threads die between one iteration and the next the percentages might get distorted, I think. Also, one thread that has e.g. a constant 5% of CPU usage can be shown as using any amount of CPU because now the percentage doesn’t depend on wall time anymore but on all other threads.

Reply:
Good point. I simplified the logic to avoid having "copies" of threads, it just uses one source of truth: nodeStats. It takes a snapshot of that, and creates the "NodeThreadInfo" list with the information, including CPU time.

Now I also separated the "CPU %" logic and moved it up to the client code to deal with the presentation logic (e.g. show as actual values or percentage).

I also changed the way we compute the % CPU by calculating the wall (all threads, not only node's). That should be more accurate.

Reply:
Maybe I’m misunderstanding what you are doing but now the client (i.e. the toadlet) has no way of knowing *when* a thread has used *any* CPU time because now *all* it knows is how much time a thread has used *in total*. And again, the only way to compare a thread’s runtime here is against all other threads but that doesn’t seem to be helpful, escpecially with longer-running threads (like e.g. the HTTP serving threads, or datastore-related threads) which at some point will probably have more runtime collected than any other thread and will completely obscure any useful information that can be taken from this.

I think that what you did two iterations ago (i.e. calculating the delta here and just exposing that) in combination with the current loop would be best.

Reply:
Ah, I don't know what I was thinking. You're right. I re-introduced the "delta" calculation although I tried to simplify it. Now DefaultThreadDiagnostics will calculate the delta CPU Time for each thread and if no delta can be calculated it will return 0.
The percentage calculation I'd leave it to the client code to deal with it.